[TVMC] Add support for the MLF to 'compile' command

gromero opened the pull request:
Hi,

Could I get reviews for the following change, please?

It adds support for the Model Library Format (MLF) to 'tvmc' so users can
output compilation artifacts to a MLF archive passing the new flag
'--output-format mlf'. For instance:

$ python3 -m tvm.driver.tvmc compile ./sine_model.tflite --target="c" --output sine.tar --output-format mlf

will generate a sine.tar archive that is serialized accordingly to the
MLF.

Since the MLF is currently meant to be used only on micro targets, an
error is generated if one tries to run a MLF outside a micro context:

$ python3 -m tvm.driver.tvmc run ./sine.tar
Error: You're trying to run a model saved using the Model Library Format (MLF). MLF can only be used to run micro targets (µTVM).

The micro context does not exist yet but will be later introduced as
part of the [RFC] "TVMC: Add support for µTVM".

Signed-off-by: Gustavo Romero <gustavo.romero@linaro.org>

gromero commented:
cc @areusch @jwfromm @mehrdadh @leandron

leandron reviewed (commented):
LGTM, thanks for doing this work @gromero.

As we spoke, you'll be adding some tests, which is great, plus, I got a couple minor comments. Feel free to use them as suggestions.

leandron commented:
Also just to report, I tested it locally and it works.

areusch commented on python/tvm/driver/tvmc/model.py:
```diff
@@ -53,6 +53,7 @@
 from tvm import relay
 from tvm.contrib import utils
 from tvm.relay.backend.executor_factory import GraphExecutorFactoryModule
+from tvm.micro import export_model_library_format
```
i'm wondering if it makes sense to keep `USE_MICRO`. i may propose removing it on Discuss. Currently, `USE_MICRO` means to build the utvm_rpc_common library from the CRT into `libtvm_runtime.so`. I don't know that adding this is a huge burden, but on the other hand I do want to be sensitive to people trying to compile for smaller runtime environments.

gromero replied:
@leandron  fwiw if `USE_MICRO OFF` an error is already generated:
```    
<snip>
raise ImportError("micro tvm is not enabled. Set USE_MICRO to ON in config.cmake")
ImportError: micro tvm is not enabled. Set USE_MICRO to ON in config.cmake
```

leandron replied:
Ack. I sent https://github.com/tlc-pack/tlcpack/pull/59 to make sure the package we generate also builds with `USE_MICRO ON`, so that it is enabled by default for users wanting to use the feature implemented in this PR.

areusch commented on python/tvm/driver/tvmc/model.py:
```diff
@@ -233,6 +232,43 @@ def export_package(
+
+        if output_format in ["so", "tar"]:
+            return self.export_classic_format(executor_factory, package_path, cross, output_format)
+        else:
```
could we be explicit here (`elif output_format == "mlf"`) rather than as a fallback?

gromero replied:
sure! done.

areusch reviewed (commented):
thanks @gromero , looks great! just a small comment

gromero commented on python/tvm/driver/tvmc/compiler.py:
```diff
@@ -40,35 +40,42 @@
+        "--output-format",
+        choices=["so", "mlf"],
+        default="so",
+        help="output format. Use 'so' for shared object or 'mlf' for Model Library Format (only for µTVM targets).",
```
Done.

gromero commented:
@leandron @areusch Thanks a lot for the reviews. @leandron I'm addressing the test case and I'll push it soon, thanks for pointing it out.

gromero commented:
Hi @areusch and @leandron PTAL. In that v2 I addressed your review comments, added 3 pytest tests to check tvmc + MLF, and rebased the code resolving the conflicts with recent changes (like those related to the addition of `--desired-layout`  and `--cross-options` flags). Thanks!

leandron reviewed (commented):
LGTM just pending a test pending the fix for one of the tests.

leandron commented on tests/python/driver/tvmc/test_mlf.py:
```diff
@@ -0,0 +1,97 @@
+    assert str(exp.value) == expected_reason, on_error
+
+
+def test_tvmc_import_package_mlf(tflite_compiled_model_mlf):
```
As this test is compiled and executed in the same machine, it will need to be protected with a `pytest.importorskip("tflite")` as well, so that it will be skipped in the CI `i386` machine, like the other in this file.

gromero replied:
Fixed. Thanks!

leandron reviewed (commented):
LGTM just pending CI. I tested it locally and I'm happy with this PR.

@areusch do you want to have another look?

gromero commented:
@tmoreau89 Hello Thierry. Could you please merge it?

leandron commented:
This is merged now. thanks @areusch @gromero!
